Closes AgileVentures/MetPlus_tracker#438

nchristian777 opened the pull request:
-Adds authentication to job applications controller.
-Adds job_application_policy.rb and corresponding test job_application_policy_spec.rb
-Adds authentication tests to job_applications_controller_spec.rb
-Refactors job_applications_controller.rb and job_applications_controller_spec.rb to comply with Rubocop standards.

joaopapereira commented on app/policies/job_application_policy.rb:
```diff
@@ -0,0 +1,20 @@
+  private
+
+  def company_person?(user)
+    User.is_company_person? user
```
I believe this validation is not enough. The Company of the user need to be the same as the company the job application is to.

Let's imagine the Job Seeker Apply to Company 1 and Company 2.
Company 2 people cannot see that the Job Seeker applied to Company 1
Company 1 people cannot see that the Job Seeker applied to Company 2

joaopapereira replied:
It should be something similar to
def company_person?(user)
  User.is_company_person? user and user.pets_user.company == resource.job.company
end

joaopapereira commented on spec/policies/job_application_policy_spec.rb:
```diff
@@ -0,0 +1,22 @@
+
+  permissions :accept?, :reject?, :show? do
+    it 'denies access if user is not a company admin/contact' do
+      expect(JobApplicationPolicy).not_to permit(job_seeker)
```
We need to add tests with the following users:
* Agency Admin
* Job Developer
* Case Manager
* Company Admin from another company
* Company Contact from another company

joaopapereira commented on spec/controllers/job_applications_controller_spec.rb:
```diff
@@ -1,69 +1,153 @@
+          it_behaves_like 'unauthenticated request'
+        end
+      end
+      context 'Job Seeker' do
```
We need to duplicate this context with the following users:
* Agency Admin
* Job Developer
* Case Manager
* Company Admin from another company
* Company Contact from another company

joaopapereira replied:
We also need to check that all the actions are tests, so we are missing

- PATCH :reject
- GET Show

patmbolger commented on app/controllers/job_applications_controller.rb:
```diff
@@ -1,54 +1,64 @@
+    else
+      flash[:notice] = 'Job application rejected.'
+      redirect_to controller: 'jobs', action: 'applications',
+                  id: @job_application.job.id unless request.xhr?
```
Not part of this story but might as well clean this up.  That is, drop this from the statement:
```
unless request.xhr?
```

nchristian777 commented:
Made the following improvements based on feedback

- job_application_policy denies company people who are from a different company
- added tests in job_application_policy_spec.rb which include the following
- - tests for agency admin, job developer, case manager
- - tests for company people form the wrong company
- added tests in job_application_controller_spec which include the following
- - tests for agency admin, job developer, case manager
- - tests for company people form the wrong company
- - tests for PATCH reject and GET show

patmbolger commented:
Changes look good to me - note that I did not check that all of João's concerns were addressed.

joaopapereira commented:
Going to merge
